art/attacks/evasion/pixel_threshold.py

```diff
@@ -0,0 +1,1346 @@
+                  Only provide this parameter if you'd like to use true labels
+                  when crafting adversarial samples. Otherwise, model
+                  predictions are used as labels to avoid the "label leaking"
+                  effect
+                  (explained in this paper: https://arxiv.org/abs/1611.01236).
+                  Default is `None`.
+        :type y: `np.ndarray`
+        :return: An array holding the adversarial examples.
+        :rtype: `np.ndarray`
+        """
+        y = check_and_transform_label_format(
+            y, self.classifier.nb_classes(), return_one_hot=False)
+
+        if y is None:
+            if self.targeted:
+                raise ValueError(
+                    'Target labels `y` need to be provided for a targeted '
+                    'attack.'
+                )
+            y = np.argmax(self.classifier.predict(x), axis=1)
+        else:
+            if len(y.shape) > 1:
+                y = np.argmax(y, axis=1)
+
+        if self.th is None:
+            logger.info('Performing minimal perturbation Attack.')
+
+        if isinstance(self.classifier, PyTorchClassifier):
+            x = np.swapaxes(x, 1, 3).astype(np.float32)
+
+        if np.max(x) <= 1:
+            x = x * 255.
+
+        adv_x_best = []
+        for image, target_class in zip(x, y):
+            if self.th is None:
+                self.min_th = 127
+                start, end = 1, 127
+                while True:
+                    image_result = []
+                    threshold = (start + end) // 2
+                    success, trial_image_result = self.attack(
+                        image, target_class, threshold)
+                    if image_result or success:
+                        image_result = trial_image_result
+                    if success:
+                        end = threshold - 1
+                    else:
+                        start = threshold + 1
+                    if success:
+                        self.min_th = threshold
+                    if end < start:
+                        break
+            else:
+                success, image_result = self.attack(
+                    image, target_class, self.th)
+            adv_x_best += [image_result]
+
+        adv_x_best = np.array(adv_x_best)
+
+        if np.max(x) <= 1:
+            x = x / 255.
+
+        if isinstance(self.classifier, PyTorchClassifier):
+            x = np.swapaxes(x, 1, 3).astype(np.float32)
+            adv_x_best = np.swapaxes(adv_x_best, 1, 3).astype(np.float32)
+
+        if y is not None:
+            y = to_categorical(y, self.classifier.nb_classes())
+
+        rate_best = 100 * \
+            compute_success(self.classifier, x, y, adv_x_best,
+                            self.targeted, 1)
+        logger.info('Success rate of Attack: %.2f%%', rate_best)
+        return adv_x_best
+
+    def predict(self, test_x):
```

Comment:
Please replace the use and definition of this function with `classifier.predict`.

Reply:
No, actually, when PyTorch is used the shape is (img_channel, img_rows, img_cols), however, my code requires the image in channels_last format. Therefore, I check if the PyTorch is used.
Please note that I change x to channels_last format and similarly the adv_x generated is in channels_last format. Therefore I need to change it again to channels_first format for the PyTorch.
Therefore to use `classifier.predict`, I need to change it according to the format required by the classifier. The function itself corrected the image format for the classifier.predict function.

Reply:
Please note that the `PyTorchClassifier` expects an input of shape `(1,28,28) or Channels First` and not `(28,28,1) or Channels Last`. Therefore, input would be in the shape `(1,28,28) or Channels First`. However, my attacks use the format `(28,28,1) or Channels Last` and these lines change it to the required format. 
This function changes the format before the `classifier.predict` is called.

Reply:
Ok, in that case please check for the value of `classifier.channel_index` to be 1 or 3 instead of checking for `isinstance(self.classifier, PyTorchClassifier)` because your attack could also be used, for example, with `KerasClassifier` and channel_index=1.

Reply:
I see, indeed checking for the `classifier.channel_index` makes more sense then checking for `class: PyTorchClassifier`. 
I have updated the checking.